packages/flutter/lib/src/material/button_style_button.dart

```diff
@@ -389,6 +389,15 @@ class _ButtonStyleState extends State<ButtonStyleButton> with TickerProviderStat
       });
     }
 
+    Color? effectiveIconColor() {
+      return widgetStyle?.iconColor?.resolve(statesController.value) ??
+          themeStyle?.iconColor?.resolve(statesController.value) ??
+          widgetStyle?.foregroundColor?.resolve(statesController.value) ??
+          themeStyle?.foregroundColor?.resolve(statesController.value) ??
+          defaultStyle.iconColor?.resolve(statesController.value) ??
+          defaultStyle.foregroundColor?.resolve(statesController.value);
```

Comment:
Can we add a comment above here to mention that in M2, iconColor doesn't have a default value? Because I was wondering why we still need to check the default foreground color if we already have default icon color😄

Reply:
Great suggestion! After looking closely, I just found out `foregroundColor` is also required  `IconButton` widget as it doesn't support `ButtonStyle.iconColor`. I can fix this in another PR. 

https://github.com/flutter/flutter/blob/bfa9a4ed3f2cff800704343c0ecda80bcefe1a7b/packages/flutter/lib/src/material/icon_button.dart#L675-L690

Reply:
Since  this would default to `foregroundColor` also in  M3 in some cases.

```dart
          defaultStyle.iconColor?.resolve(statesController.value) ??
          // Fallback to foregroundColor if iconColor is null.
          defaultStyle.foregroundColor?.resolve(statesController.value);
```

Reply:
FYI, Added `autosubmit` label as this `IconButton` discussion wouldn't block this PR.

Reply:
@TahaTesser concerning **IconButton**, what happens if you use an `IconButton` where you via theme or widget prop provide a `ButtonStyle`, where you have set `iconColor`, but not `foregroundColor`. You can do this with the "raw" `ButtonStyle`, since the `iconColor` property exists there.

Does it work logically? My expectation would be that it works like the buttons with label and icon, even if it only has an icon. We often share same "Raw" `ButtonStyle` for  buttons with label and icon and `IconButton`.

Reply:
ohhh... `IconButton` actually respects `ButtonStyle.iconColor` in M3 without `foregroundColor`. 

<img width="97" alt="Screenshot 2025-02-12 at 13 03 07" src="https://github.com/user-attachments/assets/6ba98576-d6b9-4b1f-8c5f-00ffee2df5f9" />
<img width="456" alt="Screenshot 2025-02-12 at 13 03 22" src="https://github.com/user-attachments/assets/211d71e1-1b1d-436f-8f29-62fb01126630" />

Reply:
However, this isn't available in `IconButton.styleFrom`. Only `ButtonStyle.iconColor` Is possible for `IconButtom` at the moment. This is worth aligning. 

<img width="339" alt="Screenshot 2025-02-12 at 13 05 45" src="https://github.com/user-attachments/assets/685e3aa9-50c9-4a31-ba36-50cb375a8ed3" />

Reply:
Filed https://github.com/flutter/flutter/issues/163126. Please feel free to comment, if necessary.